### PR TITLE
Random refactorings

### LINKED ISSUE #1528: remove badger.userAllow
We do random bookeeping for this thing. We only use it once, in that place using 

```javascript
if (badger.storage.getAction(domain) === constants.USER_ALLOW){
```
would be more appropriate

```bash
➜  js git:(master) ✗ grep -ri userallow
storage.js:   * userAllow list if needed
storage.js:    var index = badger.userAllow.indexOf(domain);
storage.js:      badger.userAllow.splice(index, 1);
storage.js:      badger.userAllow.push(domain);
storage.js:   * Remove user set action from a domain and remove it from userAllow
storage.js:    var index = badger.userAllow.indexOf(domain);
storage.js:      badger.userAllow.splice(index, 1);
background.js:  this.userAllow = [];
background.js:      badger.initializeUserAllowList();
background.js:  initializeUserAllowList: function() {
background.js:        this.userAllow.push(domain);
webrequest.js:    socialWidgetsToReplace[socialWidgetName] = !(badger.userAllow.indexOf(socialwidget.domain) > -1);

```

### PR DESCRIPTION
**Pull request by cowlicks**
I was in an airplane without internet for 3 hours so I did this.

closes #1528 

I'll be pulling pieces out of this into separate PR's as needed, or as I find time.

### CODE REVIEW

**Comment by koops76** [comment]
Do you want me to do more refactorings?

**Comment by cowlicks** [comment]
@koops76 yes! That would be most welcome. The code is a mess in many places. However you might want to ping me first since there are many open pull requests with refactorings that might conflict.

Things I think are very needed are mostly related to testing, I think we need to:
* make code in files exportable so we can test it, like in webrequests
* refactoring to separate code from its initialization. like we should define the `Badger` object in a different file from where we initialize it

Is there anything not mentioned that you are interested in?

**Comment by koops76** [comment]
@cowlicks What do you think about adopting [StandardJS](https://standardjs.org) code style or another widely used code style?

File: src/tests/tests/lists.js [hunk 1]
```diff
@@ -0,0 +1,12 @@
+(function() {
+  QUnit.module("Test lists");
+  let mdfp = require('multiDomainFP');
+
+  QUnit.test("Check MDFP domains are all base domains", (assert) => {
+    for (var group of mdfp.multiDomainFirstPartiesArray) {
+      for (var domain of group) {
+        assert.ok(window.getBaseDomain(domain) == domain);
```
Comment (cowlicks) [inline src/tests/tests/lists.js]: related to #1515

**Comment by koops76** [comment]
What do you think about using [Webpack](https://webpack.js.org) instead of a jerry-built implementation of CommonJS?

**Comment by cowlicks** [comment]
@koops76 that would be great. There might even be an issue for it already. Do you have a suggested code style? What are we closest too. For general style changes, I'd recommend making pull requests one file at a time. 

That would also help us move toward using strict mode (I think that has performance benefits as well).

**Comment by koops76** [comment]
Also, Webpack would allow us to use real ECMAScript modules. It would also theoretically allow us to use hot module reloading making debugging easier.

**Comment by cowlicks** [comment]
@koops76 I'm not familiar with webpack, so I can't say.

I think there are some low hanging fruit in our dependencies in build process that could be resolved pretty easily. For example we ship our test suite with the app. We should not do that. Things inside `src/lib/` are a mixture of hand edited dependencies, copied dependencies, and our own files. That is bad too.

**Comment by cowlicks** [comment]
@koops76 btw we have public meetings Monday and Thursday (tomorrow) at 11:30 pst. And have an IRC channel which you can usually find me in, #privacybadger on oftc. Feel free to drop by there and chat.

File: src/js/heuristicblocking.js [hunk 2]
```diff
@@ -533,11 +533,14 @@ function startListeners() {
...
 
-/************************************** exports */
-var exports = {};
-exports.HeuristicBlocker = HeuristicBlocker;
-exports.startListeners = startListeners;
-exports.hasCookieTracking = hasCookieTracking;
-return exports;
-/************************************** exports */
+function restart() {
+  tabOrigins = {};
```
Comment (cowlicks) [inline src/js/heuristicblocking.js]: related to #1416

**Comment by ghostwords** [comment]
I'd like to caution against refactoring for the sake of refactoring (as opposed to refactoring something as part of solving a user issue). Consider that Privacy Badger is a privacy tool with over 1.5 million users (and growing) that unfortunately provides a less-than-ideal browsing experience in many respects. While having a jerry-rigged not-a-real-module-loader-module-loader setup could be annoying to developers, is it worth spending time on over bugs that pain tens of thousands of users? Things that are fun to work on are not necessarily the things we should work on.

**Comment by koops76** [comment]
@ghostwords I disagree, code quality directly correlates with security and stability.

**Comment by ghostwords** [comment]
Sure, I'm just reminding what should matter the most: Privacy Badger's users. I'm happy to discuss the pros and cons of specific proposals.

**Comment by ghostwords** [comment]
Something else to keep in mind is that while existing Privacy Badger code may be inefficient or just plain ugly, it's code that's being used in production by 1.5m+ users. All changes to the codebase have the potential to introduce brand new bugs.

**Comment by ghostwords** [comment]
@koops76 I'm just asking you to consider your proposals in the broader context, that's all. Just take some time before disagreeing. We want the same things for Privacy Badger users. Come chat at our public meetings!

**Comment by ghostwords** [comment]
As for this PR, I would like commits related to certain issues to be part of PRs fixing those issues. For example, I think 601e75ce1029b83a28689683af247e2dd900c19f belongs with #1568, as previously requested [here](https://github.com/EFForg/privacybadger/pull/1550#issuecomment-322349990).

Having a bunch of "random" refactorings all lumped together makes it harder to figure out the importance of these refactorings, harder to recognize their connection (if any) to user-facing issues, and harder to see why changes happened when reviewing code history in the future.

**Comment by cowlicks** [comment]
@ghostwords Sorry if the intention of this was not clear, this PR (and often others' which I make) are not intended to be immediately code reviewed and merged. This also is not a WIP PR. This PR is mostly for posterity. The intention is to share the changes with other contributors, so that they can build of it, see what others are working on, etc etc.

**Comment by ghostwords** [comment]
I think we should continue the main outstanding thread from here in #1528. The other commits have either already been merged (#1591), or add more noise than signal to the repository at this point.

**Pull request closed without merging**